### PR TITLE
Cache Expended Posting on ingesters

### PR DESCRIPTION
**Pull request by alanprot**
**What this PR does**:
ThisPR introduce caching expanded postings, similar to the approach used in SG, as seen in [this change](https://github.com/thanos-io/thanos/pull/6420).

* Key Updates:
   1. Compacted Block Caching:
        * Expanded postings are cached to all queries for compacted blocks. This is safe as compacted blocks are immutable.
   1. Head Block Caching:
        * Caching is applied only to select queries containing an equal matcher for the metric name (__name__).
        For such cases, a seed is added to the cache key, allowing invalidation whenever a new series is added or removed for a given metric name.
        * Example:
           * Suppose the seed for cpu_utilization is 21. Queries would generate cache keys as follows:
              * sum(cpu_utilization{service="A"}) => Key: `21|head|__name__=cpu_utilization|service=A`
               * sum(cpu_utilization{service="A", op="GET"}) => Key: `21|head|__name__=cpu_utilization|service=A|op=GET`
    1. Cache Invalidation:
        On data ingestion, if a new series is added with `cpu_utilization` as the metric name, the seed changes, invalidating all cached entries for that metric.
    1. Memory Bound for Seeds:
        To control memory usage, a fixed-size seed slice is used. Metric names are hashed to determine their respective seed.

In load testing, this change demonstrated significant CPU and latency improvements on ingesters experiencing heavy query loads from rulers. In such cases, the same queries are evaluated constantly (every evaluation interval), and the postings are expanded for every iteration, leading to unnecessarily high CPU usage.

This are some graphs:
Screenshot 2024-11-01 at 2 40 32 PM

Flame Graphs Before and After (we can see that the after the CPU is being used mostly for the `Push` method, where before was mostly being used by the `QueryStream` method:

Screenshot 2024-11-01 at 2 44 28 PM

Screenshot 2024-11-01 at 2 42 57 PM

The idea was inspired on https://github.com/thanos-io/thanos/pull/6420 and the promise based cache implementation by https://github.com/grafana/mimir-prometheus/pull/14

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by justinjung04** [comment]
I love the test results! Thanks for doing this!!

File: pkg/storage/tsdb/expanded_postings_cache.go [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+	)
+
+	var size int
+	for _, m := range ms {
+		size += len(seed) + len(blockID.String()) + len(m.Name) + len(m.Value) + typeLen + 2*sepLen
+	}
+	sb := strings.Builder{}
+	sb.Grow(size)
+	sb.WriteString(seed)
+	sb.WriteByte('|')
```
Comment (justinjung04) [inline pkg/storage/tsdb/expanded_postings_cache.go:236]: nit: should we skip `|` if the seed is an empty string?

File: pkg/storage/tsdb/expanded_postings_cache.go [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+	h := MemHashString(metricName)
+	i := h % uint64(len(c.headSeedByMetricName))
+	l := h % uint64(len(c.strippedLock))
+	c.strippedLock[l].Lock()
+	defer c.strippedLock[l].Unlock()
+	c.headSeedByMetricName[i]++
+}
+
+func (c *BlocksPostingsForMatchersCache) PostingsForMatchers(ctx context.Context, blockID ulid.ULID, ix tsdb.IndexReader, ms ...*labels.Matcher) (index.Postings, error) {
```
Comment (justinjung04) [inline pkg/storage/tsdb/expanded_postings_cache.go:141]: Are we missing unit test for this func?

File: pkg/storage/tsdb/expanded_postings_cache.go [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+		blocksCache:             newFifoCache[[]storage.SeriesRef](cfg.Blocks, "block", metrics, cfg.timeNow),
+		headSeedByMetricName:    make([]int, seedArraySize),
+		strippedLock:            make([]sync.RWMutex, numOfSeedsStripes),
+		postingsForMatchersFunc: cfg.PostingsForMatchers,
+		timeNow:                 cfg.timeNow,
+		metrics:                 metrics,
+	}
+}
+
+func (c *BlocksPostingsForMatchersCache) ExpireSeries(metric labels.Labels) {
```
Comment (justinjung04) [inline pkg/storage/tsdb/expanded_postings_cache.go:127]: Are we missing unit test for this func?

File: docs/blocks-storage/querier.md [hunk 4]
```diff
@@ -1544,4 +1544,42 @@ blocks_storage:
...
     [enable_native_histograms: <boolean> | default = false]
+
+    # [EXPERIMENTAL] If enabled, ingesters will cache expanded postings when
+    # querying blocks. Caching can be configured separately for the head and
+    # compacted blocks.
+    expanded_postings_cache:
+      head:
+        # Max bytes for postings cache
+        # CLI flag: -blocks-storage.expanded_postings_cache.head.max-bytes
+        [max_bytes: <int> | default = 10485760]
```
[the remark this answers is not in the document]
Comment (alanprot) [reply docs/blocks-storage/querier.md:1555]: I agree! I think bytes makes more sense.

File: pkg/ingester/ingester_test.go [hunk 5]
```diff
@@ -5080,6 +5082,260 @@ func TestIngester_instanceLimitsMetrics(t *testing.T) {
...
+
+					require.Len(t, runQuery(t, ctx, i, append(m.matchers, &client.LabelMatcher{Type: client.EQUAL, Name: labels.MetricName, Value: name})), m.expectedLenght)
+
+					// first metric name should be expired
+					if in == 0 {
+						// Query only head as the block is already cached and the head was expired
+						require.Equal(t, postingsForMatchersCalls.Load(), int64(c.expectedHeadPostingCall))
+					} else {
+						require.Equal(t, postingsForMatchersCalls.Load(), int64(0))
+					}
```
[the remark this answers is not in the document]
Comment (alanprot) [reply pkg/ingester/ingester_test.go:5313]: This is testing that the cache keys from the head are invalidated upon a new series for a given metric name is added.


So, [here](https://github.com/cortexproject/cortex/pull/6296/files#diff-25e63d3b817863ee81ae61ea5bdf7dc4370cdc5033db9066d2afb1c415887d28R5297) we create a new metric with an extra label but the same metric name.. this should trigger caches for that metric name to be invalidate on the head, but not on the compacted blocks - as compacted blocks are immutable.

File: pkg/storage/tsdb/expanded_postings_cache.go (hunk 3 above)
Comment (alanprot) [reply pkg/storage/tsdb/expanded_postings_cache.go:127]: DItto.

If we break this function the tests should fail: (as this is related to cache invalidation on the head we only see the head test fail)
Screenshot 2024-11-02 at 6 42 49 PM

File: pkg/storage/tsdb/expanded_postings_cache.go (hunk 2 above)
Comment (alanprot) [reply pkg/storage/tsdb/expanded_postings_cache.go:141]: I dont think so, both `TestExpandedPostingsCacheFuzz` and `TestExpendedPostingsCache` tests this function.

File: pkg/storage/tsdb/expanded_postings_cache.go (hunk 1 above)
Comment (alanprot) [reply pkg/storage/tsdb/expanded_postings_cache.go:236]: Maybe.. I think i would avoid a condition here just for readability tbh.

File: pkg/storage/tsdb/expanded_postings_cache.go (hunk 2 above)
Comment (alanprot) [reply pkg/storage/tsdb/expanded_postings_cache.go:141]: Current Coverage:

```
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:127:		ExpireSeries					88.9%
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:141:		PostingsForMatchers				100.0%
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:145:		fetchPostings					95.8%
```

File: docs/blocks-storage/querier.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanprot) [reply docs/blocks-storage/querier.md:1555]: Done!

File: pkg/ingester/ingester_test.go [hunk 6]
```diff
@@ -5080,6 +5082,260 @@ func TestIngester_instanceLimitsMetrics(t *testing.T) {
...
+			cacheConfig: cortex_tsdb.TSDBPostingsCacheConfig{
+				Head: cortex_tsdb.PostingsCacheConfig{
+					Enabled: false,
+				},
+				Blocks: cortex_tsdb.PostingsCacheConfig{
+					Enabled: false,
+				},
+			},
+		},
+		"enabled cache blocks": {
```
[the remark this answers is not in the document]
Comment (alanprot) [reply pkg/ingester/ingester_test.go]: Done

File: docs/blocks-storage/querier.md [hunk 7]
```diff
@@ -1544,4 +1544,34 @@ blocks_storage:
...
+
+        # TTL for postings cache
+        # CLI flag: -blocks-storage.expanded_postings_cache.head.ttl
+        [ttl: <duration> | default = 10m]
+
+        # Whether the postings cache is enabled or not
+        # CLI flag: -blocks-storage.expanded_postings_cache.head.enabled
+        [enabled: <boolean> | default = false]
```
Comment (alanprot) [inline docs/blocks-storage/querier.md:1563]: TBH i think this is also redundant, i will enabled only if TTL > 0

File: docs/blocks-storage/querier.md [hunk 8]
```diff
@@ -1544,4 +1544,34 @@ blocks_storage:
     # [EXPERIMENTAL] True to enable native histogram.
     # CLI flag: -blocks-storage.tsdb.enable-native-histograms
     [enable_native_histograms: <boolean> | default = false]
+
+    # [EXPERIMENTAL] If enabled, ingesters will cache expanded postings when
+    # querying blocks. Caching can be configured separately for the head and
+    # compacted blocks.
```
Comment (yeya24) [inline docs/blocks-storage/querier.md:1550]: It is not clear to me if this cache is global or per block for both head and disk. Maybe I missed something. Do we need to document that better?

File: docs/blocks-storage/querier.md (hunk 7 above)
Comment (yeya24) [reply docs/blocks-storage/querier.md]: There is no usecase to have 0 TTL so I never evict based on time but evict based on cache size? This is what we do for our inmemory index cache

Comment (alanprot) [reply docs/blocks-storage/querier.md]: humm... fair... even though its not a very common use case...

I can put it back! hahha

File: docs/blocks-storage/querier.md (hunk 8 above)
Comment (alanprot) [reply docs/blocks-storage/querier.md:1550]: fair.. added it back! :P

File: docs/blocks-storage/querier.md (hunk 7 above)
Comment (alanprot) [reply docs/blocks-storage/querier.md:1568]: added it back

File: docs/blocks-storage/querier.md (hunk 8 above)
Comment (yeya24) [reply docs/blocks-storage/querier.md:1550]: Yes. Looks like it is global?

Comment (alanprot) [reply docs/blocks-storage/querier.md:1550]: Yeah.. its global.

We create a cache for head and for the disk. And the disk one is "global". 

Lemme try to clarify a bit.

Comment (alanprot) [reply docs/blocks-storage/querier.md:1550]: Done!

File: pkg/ingester/ingester.go [hunk 9]
```diff
@@ -2211,6 +2229,12 @@ func (i *Ingester) createTSDB(userID string) (*userTSDB, error) {
 		OutOfOrderCapMax:               i.cfg.BlocksStorageConfig.TSDB.OutOfOrderCapMax,
 		EnableOverlappingCompaction:    false, // Always let compactors handle overlapped blocks, e.g. OOO blocks.
 		EnableNativeHistograms:         i.cfg.BlocksStorageConfig.TSDB.EnableNativeHistograms,
+		BlockChunkQuerierFunc: func(b tsdb.BlockReader, mint, maxt int64) (storage.ChunkQuerier, error) {
```
Comment (yeya24) [inline pkg/ingester/ingester.go:2232]: This hook seems doesn't work for the new Head and OOO querier as they bypass the hook and create their own querier instead.
https://github.com/prometheus/prometheus/blob/main/tsdb/db.go#L2177

Not a blocker for this PR.

**Review by yeya24 — APPROVED** [review]

File: pkg/ingester/ingester.go (hunk 9 above)
Comment (alanprot) [reply pkg/ingester/ingester.go:2232]: i will create a issue after to track.

**Pull request merged**